### PR TITLE
Refactor for protein sequences

### PR DESCRIPTION
**Pull request by iskandr**
Another step on the road to accepting inputs such as 'EGFR T790M', I deleted a lot of the custom objects which assume that we always have access to a genomic variant (e.g. `MutantEpitopePrediction`) and made `TopiaryPredictor` methods for making predictions on protein sequences in the absence of variants. 

Results are now returned in DataFrames, whose columns are documented in function docstrings. This isn't great for maintainability, but neither was having a sprawling mass of custom data objects. 

I also deleted `LazyLigandomeDict` since we never use it in practice and propagating it through new code made refactoring even harder. In the future we should port over the reference proteome index from vaxrank.

### CODE REVIEW

File: topiary/predictor.py [hunk 1]
```diff
@@ -159,26 +228,123 @@ def epitopes_from_mutation_effects(
...
+        # TODO: add extra boolean field
+        #   novel = is_mutant | not_in_reference
+        # Requires keeping a quick lookup structure for all peptides in
+        # the reference proteome
+        if self.only_novel_epitopes:
+            df = df[df.contains_mutant_residues]
+
+        return df
+
+    def predict_variants(
```
Comment (tavinathanson) [inline topiary/predictor.py]: Weird name, sounds like you're predicting variants. Maybe `predict_from_variants`?

File: topiary/predictor.py [hunk 2]
```diff
@@ -211,7 +401,7 @@ def epitopes_from_variants(
 
         effects = variants.effects(raise_on_error=self.raise_on_error)
 
-        return self.epitopes_from_mutation_effects(
+        return self.predict_mutation_effects(
```
Comment (tavinathanson) [inline topiary/predictor.py]: `predict_from_mutation_effects`?

File: topiary/cli/script.py [hunk 3]
```diff
@@ -54,12 +50,6 @@ def main(args_list=None):
     args = parse_args(args_list)
     print("Topiary commandline arguments:")
     print(args)
-    epitopes = predict_epitopes_from_args(args)
```
Comment (tavinathanson) [inline topiary/cli/script.py]: Am I understanding this right, that `predict_epitopes_from_args` used to grab `gene_expression_dict` internally, but we also needed to grab `gene_expression_dict` separately in the next line to pass it through to `epitopes_to_dataframe`?

Why would that have been necessary if the epitopes were already filtered in `predict_epitopes_from_args`? And if they weren't filtered, how will `df = predict_epitopes_from_args(args)` work?

File: topiary/predictor.py [hunk 4]
```diff
@@ -91,10 +86,60 @@ def __init__(
...
+            - source_sequence_name
+            - peptide
+            - peptide_offset
+            - peptide_length
+            - allele
+            - affinity
+            - percentile_rank
+            - prediction_method_name
+        """
+        df = self.mhc_model.predict_subsequences_dataframe(name_to_sequence_dict)
```
Comment (tavinathanson) [inline topiary/predictor.py]: Confused about where named vs. non-named existed prior?

File: topiary/predictor.py [hunk 5]
```diff
@@ -159,26 +228,123 @@ def epitopes_from_mutation_effects(
...
+        logging.info("MHC predictor returned %d peptide binding predictions" % (
+            len(df)))
+
+        # since we used variant descrptions as the name of each sequence
+        # let's rename that column to be more informative
+        df = df.rename(columns={"source_sequence_name": "variant"})
+
+        # adjust offset to be relative to start of protein, rather
+        # than whatever subsequence we used for prediction
+        def compute_peptide_offset_relative_to_protein(row):
```
Comment (tavinathanson) [inline topiary/predictor.py]: Using `peptide_offset` here to then generate a `peptide_offset` seems error-prone. Can you call the new `peptide_offset` by a new name?

File: topiary/predictor.py [hunk 6]
```diff
@@ -159,26 +228,123 @@ def epitopes_from_mutation_effects(
...
-            epitopes,
-            ic50_cutoff=self.ic50_cutoff,
-            percentile_cutoff=self.percentile_cutoff,
-            only_novel_epitopes=self.only_novel_epitopes)
-
-    def epitopes_from_variants(
+        # since we know that each set of variant effects has been
+        # reduced to a single 'top priority' effect, we can uniquely
+        # identify each variant sequence by its original genomic variant
+        variant_string_to_effect_dict = {
```
Comment (tavinathanson) [inline topiary/predictor.py]: File an issue to update this?

File: topiary/predictor.py [hunk 7]
```diff
@@ -159,26 +228,123 @@ def epitopes_from_mutation_effects(
...
+            effect.variant.short_description: effect
+            for effect in effect_to_subsequence_dict.keys()
+        }
+        variant_string_to_subsequence_dict = {
+            effect.variant.short_description: seq
```
Comment (tavinathanson) [inline topiary/predictor.py]: Nit: call `seq` `subseq`?

Comment (iskandr) [reply topiary/predictor.py]: 👍

File: topiary/predictor.py (hunk 5 above)
Comment (iskandr) [reply topiary/predictor.py]: I keep making that field longer, it was just `offset` before...

How about `peptide_offset_in_protein`?

Comment (tavinathanson) [reply topiary/predictor.py]: 👍

**Pull request merged**